framework/src/mesh/MooseMesh.C

```diff
@@ -600,51 +600,51 @@ MooseMesh::buildNodeList()
 {
   freeBndNodes();
 
-  /// Boundary node list (node ids and corresponding side-set ids, arrays always have the same length)
-  std::vector<dof_id_type> nodes;
-  std::vector<boundary_id_type> ids;
-  getMesh().get_boundary_info().build_node_list(nodes, ids);
-
-  int n = nodes.size();
-  _bnd_nodes.resize(n);
-  for (int i = 0; i < n; i++)
+  auto bc_tuples = getMesh().get_boundary_info().build_node_list();
+
+  int n = bc_tuples.size();
+  _bnd_nodes.clear();
+  _bnd_nodes.reserve(n);
+  for (const auto & t : bc_tuples)
   {
-    _bnd_nodes[i] = new BndNode(getMesh().node_ptr(nodes[i]), ids[i]);
-    _node_set_nodes[ids[i]].push_back(nodes[i]);
-    _bnd_node_ids[ids[i]].insert(nodes[i]);
+    auto node_id = std::get<0>(t);
+    auto bc_id = std::get<1>(t);
+
+    _bnd_nodes.push_back(new BndNode(getMesh().node_ptr(node_id), bc_id));
```

Comment:
I thought about this, as it would allow us to drop the manual `delete`s from `freeBndNodes()` and `freeBndElems()`. There should be effectively no additional overhead for `unique_ptr` since we don't provide a custom deleter, so it would be worth looking at.